File: python-examples/stagehand/api/get-books.py
from typing import TypedDict
import os
from intuned_runtime import attempt_store, get_ai_gateway_config
from playwright.async_api import Page
from pydantic import BaseModel

from stagehand import AsyncStagehand


class Params(TypedDict):
    category: str | None


class BookDetails(BaseModel):
    title: str
    price: str
    rating: str | None = None
    availability: str | None = None


class BooksResponse(BaseModel):
    books: list[BookDetails]


MAX_PAGES = 10


async def automation(page: Page, params: Params, **_kwargs):
    base_url, api_key = get_ai_gateway_config()
    model_api_key = os.getenv("MODEL_API_KEY")
    cdp_url = attempt_store.get("cdp_url")
    print(f"CDP URL: {cdp_url}")
    print(f"Base URL: {base_url}")
    # Initialize Stagehand with act/extract/observe capabilities
    client = AsyncStagehand(
        server="local",
        model_api_key=model_api_key,
        local_ready_timeout_s=30.0,
        base_url=base_url,
    )
    print("⏳ Starting local session (this will start the embedded SEA binary)...")
    session = await client.sessions.start(
        model_name="openai/gpt-5-mini",
        browser={
            "type": "local",
            "launchOptions": {
                "headless": False,
                "cdpUrl": cdp_url,
            },
        },
    )
    session_id = session.data.session_id
    print(f"✅ Session started: {session_id}")
    print("\nInitialized 🤘 Stagehand")

    await page.set_viewport_size({"width": 1280, "height": 800})

    category = params.get("category")
    all_books: list[BookDetails] = []

    try:
        # Navigate using Stagehand's API
        await client.sessions.navigate(
            id=session_id,
            url="https://books.toscrape.com",
        )

        # Navigate to category if specified using act
        if category:
            # Use observe to find the category link
            observed = await client.sessions.observe(
                id=session_id,
                instruction=f'the "{category}" category link in the sidebar',
            )
            print(f"Observed category link: {observed.data.result}")

            # Use act to click on the category
            await client.sessions.act(
                id=session_id,
                input=f'Click on the "{category}" category link in the sidebar',
            )
            print(f"Navigated to {category} category")

        # Collect books from all pages
        for page_num in range(1, MAX_PAGES + 1):
            print(f"Extracting books from page {page_num}...")

            # Extract all book details from the current page using simple JSON schema
            result = await client.sessions.extract(
                id=session_id,
                instruction="Extract all books on this page including title, price, rating and availability for each book",
                schema={
                    "type": "object",
                    "properties": {
                        "books": {
                            "type": "array",
                            "items": {
                                "type": "object",
                                "properties": {
                                    "title": {"type": "string"},
                                    "price": {"type": "string"},
                                    "rating": {"type": "string"},
                                    "availability": {"type": "string"},
                                },
                                "required": ["title", "price"],
                            },
                        }
                    },
                    "required": ["books"],
                },
            )

            # Parse the result into BooksResponse
            books_data = BooksResponse.model_validate(result.data.result)
            all_books.extend(books_data.books)
            print(
                f"Found {len(books_data.books)} books on page {page_num}, total: {len(all_books)}"
            )

            # Check if there's a next page and navigate to it
            if page_num < MAX_PAGES:
                try:
                    # Use observe to check if next button exists
                    next_button = await client.sessions.observe(
                        id=session_id,
                        instruction='the "next" button or link to go to the next page',
                    )
                    if not next_button.data.result:
                        print("No more pages available")
                        break

                    # Use act to click the next button
                    await client.sessions.act(
                        id=session_id,
                        input='Click the "next" button to go to the next page',
                    )
                    print(f"Navigated to page {page_num + 1}")
                except Exception as e:
                    print(f"No more pages or navigation failed: {e}")
                    break
    finally:
        # Cleanup Stagehand
        print("\nClosing 🤘 Stagehand...")
        await client.sessions.end(session_id)
        print("✅ Session ended")

    return BooksResponse(books=all_books)

File: typescript-examples/stagehand/api/get-books.ts
import z from "zod";
import { Stagehand, AISdkClient } from "@browserbasehq/stagehand";
import { createOpenAI } from "@ai-sdk/openai";
import type { Page, BrowserContext } from "playwright";
import { attemptStore, getAiGatewayConfig } from "@intuned/runtime";


interface Params {
  category?: string;
}

const bookDetailsSchema = z.object({
  books: z.array(
    z.object({
      title: z.string(),
      price: z.string(),
      rating: z.string().optional(),
      availability: z.string().optional(),
    })
  ),
});

type BookDetails = z.infer<typeof bookDetailsSchema>["books"][number];
type BooksResponse = z.infer<typeof bookDetailsSchema>;

const MAX_PAGES = 10;

async function getWebSocketUrl(cdpUrl: string): Promise<string> {
  const versionUrl = cdpUrl.endsWith("/")
    ? `${cdpUrl}json/version`
    : `${cdpUrl}/json/version`;
  const response = await fetch(versionUrl);
  const data = await response.json();
  return data.webSocketDebuggerUrl;
}

export default async function handler(
  { category }: Params,
  page: Page,
  _context: BrowserContext
) {
  // Get AI gateway config for Stagehand
  const { baseUrl, apiKey } = await getAiGatewayConfig();
  const cdpUrl = attemptStore.get("cdpUrl") as string;
  const webSocketUrl = await getWebSocketUrl(cdpUrl);

  // Create AI SDK provider with Intuned's AI gateway
  const openai = createOpenAI({
    apiKey,
    baseURL: baseUrl,
  });

  const llmClient = new AISdkClient({
    model: openai("gpt-5-mini"),
  });

  // Initialize Stagehand with act/extract/observe capabilities
  const stagehand = new Stagehand({
    env: "LOCAL",
    localBrowserLaunchOptions: {
      cdpUrl: webSocketUrl,
      viewport: { width: 1280, height: 800 },
    },
    llmClient,
    logger: console.log,
  });
  await stagehand.init();
  console.log("\nInitialized 🤘 Stagehand");

  await page.setViewportSize({ width: 1280, height: 800 });

  const allBooks: BookDetails[] = [];

  try {
    await page.goto("https://books.toscrape.com");

    // Navigate to category if specified using act
    if (category) {
      // Use observe to find the category link in the sidebar
      const observed = await stagehand.observe(
        `the "${category}" category link in the sidebar`
      );
      console.log(`Observed category link: ${JSON.stringify(observed)}`);

      // Use act to click on the category
      await stagehand.act(
        `Click on the "${category}" category link in the sidebar`
      );
      console.log(`Navigated to ${category} category`);
    }

    // Collect books from all pages
    for (let pageNum = 1; pageNum <= MAX_PAGES; pageNum++) {
      console.log(`Extracting books from page ${pageNum}...`);

      // Extract all book details from the current page
      const result = await stagehand.extract(
        "Extract all books visible on the page with their complete details including title, price, rating, and availability",
         bookDetailsSchema ,
      );
      allBooks.push(...result.books ?? []);
      console.log(
        `Found ${result?.books?.length ?? 0} books on page ${pageNum}, total: ${allBooks.length}`
      );

      // Check if there's a next page and navigate to it
      if (pageNum < MAX_PAGES) {
        try {
          // Use observe to check if next button exists
          const nextButton = await stagehand.observe(
            'the "next" button or link to go to the next page'
          );
          if (!nextButton || nextButton.length === 0) {
            console.log("No more pages available");
            break;
          }

          // Use act to click the next button
          await stagehand.act(
            'Click the "next" button to go to the next page'
          );
          console.log(`Navigated to page ${pageNum + 1}`);
        } catch (e) {
          console.log(`No more pages or navigation failed: ${e}`);
          break;
        }
      }
    }
  } finally {
    // Cleanup Stagehand
    console.log("\nClosing 🤘 Stagehand...");
    await stagehand.close();
  }

  return { books: allBooks } satisfies BooksResponse;
}
